### DISCUSSION TITLE
Overriding default typeRoots does not seem to be working

### DISCUSSION THREAD

**Discussion by hanselabreu** [nrwl/nx#4929] [General]
Does anyone have any suggestions as a way to create a `types` library that will be used to declare modules and then be referenced in other libraries? I have tried the following:

Overriding the `DefaultTheme` in the `styled-components/native` module
```javascript
// libs/types/styled.d.ts

declare module "styled-components/native" {
  export type DefaultTheme = {
    colors: string
  }
}
```

Then configuring `typeRoots` on the `libs/components` library like so:

```javascript
// libs/components/tsconfig.json

{
  ...
  "typeRoots": ["../types"]
  ...
}
```

Finally importing it on that library and getting it’s overridden values

```javascript
// libs/components

import { DefaultTheme } from "styled-components/native"
```
Where `DefaultTheme` would be an `object` with the `colors` key as a `string`, but it's not. It keeps defaulting to the `styled-components/native` interface.

Another thing, as per TypeScript's [typeRoots documentation](https://www.typescriptlang.org/tsconfig#typeRoots), if `typeRoots` is specified, only packages under `typeRoots` will be included. For example:

```javascript
{
  "compilerOptions": {
    "typeRoots": ["./typings", "./vendor/types"]
  }
}
```

This config file will include all packages under `./typings` and `./vendor/types`, and no packages from `./node_modules/@types`. All paths are relative to the `tsconfig.json`.

But this does not appear to happen either, if I have `typeRoots` specified in a library, the `./node_modules/@types` are still being defined. 

Any suggestions?

**Comment by tommyc38** [comment]
I too have had issues on how to implement typings in my Nx monorepo, particularly types that are going to be shared between a number of libraries. 

CURRENT STATE:
Create one lib per "type grouping".  This seems like overkill and could get out of control fast with many libs.

FUTURE STATE
Nx has a cli command for creating a special library for typing data (e.g.libs/shared/types <---lib).  First, you create the lib and then you create your "type grouping" which would create a typeRoot for all typing data placed in that sub folder (e.g.libs/shared/types/media <---media typeRoot (e.g. @shared-types-media).  Moreover, you could then create these special typings libs to indicate and limit scope (e.g. globally shared, app specific shared, etc.).  Lastly, when it came time to build, the incremental builder would only look to see if anything changed in the respective rootTypes folder and only build ones that changed accordingly.

**Reply by tommyc38** [reply]
This could also be the same case with shared functions

**Reply by hanselabreu** [reply]
Excellent suggestion. Would love to see something like this implemented. As it is right now, you will find yourself repeating a lot of the typings among different libraries.

**Comment by ptrhck** [comment]
I do not fully understand your solution. I have defined two modules in two seperate .d.ts files. How can I share them accross libs? I have tried multiple configs, but nothing worked. Would really appreciate help!

**Comment by aitorllj93** [comment]
something new about this? I have the same issue. I tried all the possible  combinations:

`` tsconfig.base.json``
```json
    "typeRoots": ["./node_modules",  "./typings", "libs/web-components/types"],
```

`` libs/mylib/tsconfig.json``
```json
    "typeRoots": ["../../node_modules", "./types", "../../typings"]
```

`` apps/myapp/tsconfig.app.json``
```json
    "typeRoots": [
      "../../node_modules",
      "../../libs/web-components/types",
      "../../typings"
    ]
```

**Comment by GentileFulvio** [comment]
Any news on this topic ? I seem to have the same issue

**Comment by peter-wd-1** [comment]
How about using a generator adding `"include": [../../global-types/**/*.ts]` for libs and apps tsconfig.app/lib.json on creation of libs/apps. You can add overing types there. I solved this way in my case.

**Comment by ryanflowers** [comment]
I also solved this by adding the path to include as @peter-wd-1 did. However this is a brittle solution as you now have to override the nx generators(yay) plus if you ever move your folder structure it all breaks. Just seems tsc would have a solution for this.

**Comment by shady-cansultant** [comment]
Hello from 2025  👋🏻. Any movements on this?

**Comment by viters** [comment]
**UPDATE: This solution messes release process, see comments for different approach.**

I feel this is silly, but I've tried a number of different solutions.  In other libraries, I have a dependency `internal-js-package`, but it does not provide types. I wanted to have separate typings package for it, let's call it `@types/internal-js-package`. I have created a library project for it. What finally made it recognizable from other parts of project was adding `paths` alias to `tsconfig.base.json` in root dir. It is properly recognized by editors (JetBrains and VS Code) and also is being built correctly:
```json
{
  "compilerOptions": {
    ...
    "paths": {
      "// types": ["I have no idea why the types must be aliased this way, but it works both for IDE and build"],
      "internal-js-package": ["libs/internal-js-package-types"]
    }
  },
}
```
I have `libs/internal-js-package-types/index.d.ts` holding all types, and accompanying package.json:
```json
{
  "name": "@types/internal-js-package",
  "version": "0.0.1",
  "types": "./index.d.ts",
  "dependencies": {
    "tslib": "^2.3.0"
  }
}
```
(this package does not export any code, only types)

**Reply by viters** [reply]
Solution with `"paths"` aliases indeed work for types, but messes the release process creating links between projects that should not exist.

**Reply by viters** [reply]
I have found another solution, which seems to work in all use cases. If you have `the-untyped-package` and separate types package for it `@types/the-untyped-package` containing .d.ts files, add it as dev dependency in root package.json and link it to local file:
```json
{
  "devDependencies": {
    "@types/the-untyped-package": "./libs/the-untyped-package-types"
  }
}
```
I am not sure if it will work if `the-untyped-package-types` would declare types for modules other than `the-untyped-package`.
